serde_derive/src/de.rs

```diff
@@ -1313,6 +1313,56 @@ fn prepare_enum_variant_enum(
     (variants_stmt, variant_visitor)
 }
 
+fn prepare_adjacently_tagged_enum_variant_enum(tag: &str) -> Fragment {
+    let expecting = format!("enum {}", tag);
+    quote_block! {
+        #[doc(hidden)]
+        struct __FieldVariant(pub __Field);
```

Comment:
I would expect this whole type to go in `serde::__private::de` instead of emitting a whole new type and set of trait impls into every single use of adjacently tagged enums.

```rust
pub struct AdjacentlyTaggedEnumVariant<F> {
    tag: &'static str,
    variants: &'static [&'static str],
    fields_enum: PhantomData<F>,
}

impl<'de, F> DeserializeSeed<'de> for AdjacentlyTaggedEnumVariant<F>
where
    F: Deserialize<'de>,
{
    type Value = F;
    ...
}
```

Reply:
that's what I wanted to do in the first place, but I just looked at `Deserialize` and because it don't take `self` I just moved on  and did not thought of DeserializeSeed, I did something like this for serialization